gdal/port/cpl_vsil_hdfs.cpp

```diff
@@ -0,0 +1,320 @@
+  case SEEK_SET:
+    return hdfsSeek(this->poFilesystem, this->poFile, nOffset);
+  case SEEK_CUR:
+    return hdfsSeek(this->poFilesystem,
+                    this->poFile,
+                    nOffset + this->Tell());
+  case SEEK_END:
+    return hdfsSeek(this->poFilesystem,
+                    this->poFile,
+                    static_cast<tOffset>(this->Length()) - nOffset);
+  default:
+    return -1;
+  }
+}
+
+vsi_l_offset VSIHdfsHandle::Tell()
+{
+  return hdfsTell(this->poFilesystem, this->poFile);
+}
+
+size_t VSIHdfsHandle::Read(void *pBuffer, size_t nSize, size_t nMemb)
+{
+  return hdfsRead(this->poFilesystem, this->poFile, pBuffer, nSize * nMemb);
+}
+
+size_t VSIHdfsHandle::Write(const void *pBuffer, size_t nSize, size_t nMemb)
+{
+  CPLError(CE_Failure, CPLE_AppDefined, "HDFS driver is read-only");
+  return -1;
+}
+
+int VSIHdfsHandle::Eof()
+{
+  return (this->Tell() == this->Length());
+}
+
+int VSIHdfsHandle::Flush()
+{
+  return hdfsFlush(this->poFilesystem, this->poFile);
+}
+
+int VSIHdfsHandle::Close()
+{
+  int retval = 0;
+
+  if (this->poFilesystem != nullptr && this->poFile != nullptr)
+    retval = hdfsCloseFile(this->poFilesystem, this->poFile);
+  this->poFile = nullptr;
+  this->poFilesystem = nullptr;
+
+  return retval;
+}
+
+class VSIHdfsFilesystemHandler final : public VSIFilesystemHandler
+{
+  private:
+    CPL_DISALLOW_COPY_ASSIGN(VSIHdfsFilesystemHandler)
+
+    hdfsFS poFilesystem;
+
+  public:
+    VSIHdfsFilesystemHandler();
+    ~VSIHdfsFilesystemHandler() override;
+
+    VSIVirtualHandle *Open(const char *pszFilename,
+                           const char *pszAccess,
+                           bool bSetError ) override;
+    int Stat(const char *pszFilename,
+             VSIStatBufL *pStatBuf,
+             int nFlags) override;
+    int Unlink(const char *pszFilename) override;
+    int Mkdir(const char *pszDirname, long nMode) override;
+    int Rmdir(const char *pszDirname) override;
+    char ** ReadDir(const char *pszDirname) override;
+    int Rename(const char *oldpath, const char *newpath) override;
+
+};
+
+VSIHdfsFilesystemHandler::VSIHdfsFilesystemHandler()
+{
+  this->poFilesystem = hdfsConnect("default", 0);
+}
+
+VSIHdfsFilesystemHandler::~VSIHdfsFilesystemHandler()
+{
+  if (this->poFilesystem != nullptr)
+    hdfsDisconnect(this->poFilesystem);
+  this->poFilesystem = nullptr;
+}
+
+VSIVirtualHandle *
+VSIHdfsFilesystemHandler::Open( const char *pszFilename,
+                                const char *pszAccess,
+                                bool)
+{
+  if (strchr(pszAccess, 'w') != nullptr || strchr(pszAccess, 'a') != nullptr) {
+    CPLError(CE_Failure, CPLE_AppDefined, "HDFS driver is read-only");
+    return nullptr;
+  }
+
+  if (strncmp(pszFilename, VSIHDFS, strlen(VSIHDFS)) != 0) {
+    return nullptr;
+  }
+  else {
+    const char * pszPath = pszFilename + strlen(VSIHDFS);
+    hdfsFile poFile = hdfsOpenFile(poFilesystem, pszPath, O_RDONLY, 0, 0, 0);
```

Comment:
I assume hdfs API is thread-safe ?

Reply:
Yes, the claim can be found [here](https://hadoop.apache.org/docs/r2.7.6/hadoop-project-dist/hadoop-hdfs/LibHdfs.html#Thread_Safe).  The exact wording used there seems to leave a little bit of room for a negative answer if there is a deficiency within the JNI functionality, but that is not believed to be the case.